tests/lib/nrf_modem_lib/nrf_modem_lib_trace/src/main.c

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include <unity.h>
+
+#include <zephyr/kernel.h>
+
+#include "nrf_modem_lib_trace.h"
+
+#include "mock_trace_backend.h"
+#include "mock_nrf_modem.h"
+
+extern int unity_main(void);
+
+static const char *start_trace_at_cmd_fmt = "AT%%XMODEMTRACE=1,%hu";
+static unsigned int exp_trace_mode;
+static int nrf_modem_at_printf_retval;
+
+static K_HEAP_DEFINE(trace_heap, 128);
+
+/* Suite teardown shall finalize with mandatory call to generic_suiteTearDown. */
+extern int generic_suiteTearDown(int num_failures);
+
+void setUp(void)
+{
+	mock_nrf_modem_Init();
+	mock_trace_backend_Init();
+}
+
+void tearDown(void)
+{
+	mock_nrf_modem_Verify();
+	mock_trace_backend_Verify();
+}
+
+static void nrf_modem_at_printf_ExpectTraceModeAndReturn(unsigned int trace_mode, int retval)
+{
+	exp_trace_mode = trace_mode;
+	nrf_modem_at_printf_retval = retval;
+}
+
+/* Custom mock written for nrf_modem_at_printf since CMock can not be made to test parameters
+ * of variadic functions.
+ */
+int nrf_modem_at_printf(const char *fmt, ...)
+{
+	va_list args;
+	unsigned int trace_mode;
+
+	TEST_ASSERT_EQUAL_STRING(start_trace_at_cmd_fmt, fmt);
+
+	va_start(args, fmt);
+	trace_mode = va_arg(args, unsigned int);
+	va_end(args);
+
+	TEST_ASSERT_EQUAL(exp_trace_mode, trace_mode);
+
+	return nrf_modem_at_printf_retval;
+}
+
+int test_suiteTearDown(int num_failures)
+{
+	return generic_suiteTearDown(num_failures);
+}
+
+static const uint8_t *trace_processed_callback_buf_expect;
+static uint32_t trace_processed_callback_len_expect;
+static int trace_processed_callback_retval;
+
+static int trace_processed_callback_stub(const uint8_t *buf, uint32_t len, int cmock_calls)
+{
+	TEST_ASSERT_EQUAL(buf, trace_processed_callback_buf_expect);
+	TEST_ASSERT_EQUAL(len, trace_processed_callback_len_expect);
+	return trace_processed_callback_retval;
+}
+
+/* Custom mock written for nrf_modem_trace_processed_callback since the regular mock
+ * function, __wrap_nrf_modem_trace_processed_callback_ExpectAndReturn, erroneously
+ * calls the fault handler when compiler is optimizing for size.
+ */
+static void trace_processed_callback_ExpectAndReturn(const uint8_t *buf, uint32_t len, int retval)
+{
+	trace_processed_callback_buf_expect = buf;
+	trace_processed_callback_len_expect = len;
+	trace_processed_callback_retval = retval;
+	__wrap_nrf_modem_trace_processed_callback_Stub(trace_processed_callback_stub);
+}
+
+/* Test that nrf_modem_lib_trace_start returns error when modem trace module is not initialized. */
+void test_modem_trace_start_when_not_initialized(void)
```

Comment:
@MirkoCovizzi Please update the test names to match the function and case they are testing. E.g.
```suggestion
void test_nrf_modem_lib_trace_start_eperm(void)
```